File: visionflow/backend/backend/schemas/workflow.py
"""Workflow request/response schemas."""

from __future__ import annotations

from pydantic import BaseModel, Field

from shared.enums import SpeedMode
from shared.step_models import Workflow, WorkflowStep, WorkflowSummary


class WorkflowListItem(BaseModel):
    workflow_id: str
    name: str
    description: str
    version: int
    last_successful_version: int | None = None
    step_count: int = 0
    last_run_at: str | None = None
    success_rate: float | None = None


class WorkflowListResponse(BaseModel):
    workflows: list[WorkflowListItem]
    total: int


class GenerateWorkflowRequest(BaseModel):
    session_id: str


class UpdateWorkflowRequest(BaseModel):
    name: str | None = None
    description: str | None = None
    default_speed: SpeedMode | None = None
    steps: list[WorkflowStep] | None = None

File: visionflow/frontend/src/types/workflow.ts
export type StepType =
  | 'cmd'
  | 'hotkey'
  | 'clipboard_paste'
  | 'vision_click'
  | 'navigate'
  | 'scroll'
  | 'drag'
  | 'wait'
  | 'file_open'
  | 'file_write'
  | 'focus_window';

export type OnFailure = 'retry' | 'human' | 'skip' | 'abort';
export type SpeedMode = 'fast' | 'normal' | 'slow';

export interface WorkflowStep {
  id: number;
  type: StepType;
  description: string;
  value?: string | null;
  target_description?: string | null;
  hint?: string | null;
  wait_condition?: string | null;
  timeout_sec: number;
  speed: SpeedMode;
  on_failure: OnFailure;
  screenshot_ref?: string | null;
  crop_ref?: string | null;
}

export interface Workflow {
  workflow_id: string;
  name: string;
  description: string;
  version: number;
  last_successful_version?: number | null;
  default_speed: SpeedMode;
  steps: WorkflowStep[];
}

export interface WorkflowSummary {
  flow_summary: string;
  attention_steps: string[];
  improvement_suggestions: string[];
  recommended_first_run: 'dryrun' | 'manual' | 'auto';
  estimated_duration_sec?: number | null;
}

export interface GenerateWorkflowResponse {
  workflow: Workflow;
  summary: WorkflowSummary | null;
  generation_method: 'ai' | 'rule_based';
}

export interface WorkflowListItem {
  workflow_id: string;
  name: string;
  description: string;
  version: number;
  last_successful_version?: number | null;
  step_count: number;
  last_run_at?: string | null;
  success_rate?: number | null;
}
